source-mongodb/pull.go

```diff
@@ -3,24 +3,26 @@ package main
 import (
 	"context"
 	"encoding/json"
-	"errors"
 	"fmt"
-	"io"
 	"math"
 	"time"
 
 	boilerplate "github.com/estuary/connectors/source-boilerplate"
 	pc "github.com/estuary/flow/go/protocols/capture"
 	pf "github.com/estuary/flow/go/protocols/flow"
 	log "github.com/sirupsen/logrus"
-	"golang.org/x/sync/errgroup"
 
+	"golang.org/x/sync/errgroup"
 	"go.mongodb.org/mongo-driver/bson"
 	"go.mongodb.org/mongo-driver/bson/primitive"
 	"go.mongodb.org/mongo-driver/mongo"
 	"go.mongodb.org/mongo-driver/mongo/options"
 )
 
+// Limit the number of backfills running concurrently to avoid
+// too much memory pressure
+const ConcurrentBackfillLimit = 20
```

Comment:
If a capture has more than 20 bindings, I believe this will mean that bindings 20+ will need to wait until at least some of those initial 20 have finished.

Here's an idea: I see that there is a method `cursor.RemainingBatchLength()`, which IIUC describes how many documents have been loaded into memory for the iterator, and once that gets to `0` the next call will require a network request. With this information we may be able to efficiently process an entire batch's documents for a binding, and then potentially switch to a different binding and process a batch for it. A single threaded version of this would loop, and in each loop pick a binding that still needs a backfill done at random and process 1 batch for it. A multi-threaded version could start a separate go routine for each binding, but use a semaphore to limit the active batches to something smaller like 5 at a time, while still progressing the backfill for each binding somewhat evenly, at random.

This may work poorly though if there is a lot of overhead in requesting a new cursor vs. continuing a previous cursor. But if that's the case, I would wonder if doing 20 at a time is too many.

Reply:
@williamhbaker this actually is a very neat way of handling this. I implemented it and tested it.

I'm currently using 20 as the limit for bindings running concurrently. Given that a batch can't be larger than 16mb, it means at maximum this process will use around 320mb memory (not counting for other memory usage of the driver).

Reply:
Nice! 20 does _seem_ like a lot, although I certainly haven't tested it. Generally I haven't seen much benefit to more than 3-5 concurrent requests like these. 320mb of raw data + overhead is kind of a lot too so it might make sense to dial that back. Hard to testing meaningfully though without production use cases though.